api/posts.go

```diff
@@ -39,26 +51,120 @@ func (api API) GetPost(ctx context.Context, id uint32) (*model.Post, error) {
 	if err != nil {
 		return nil, NewError(err)
 	}
+	// default ImagesStatus, but see the comment about ImagesLoading in model/post.go
+	if post.ImagesStatus == "" {
+		post.ImagesStatus = model.PostDraft
+	}
 	return post, nil
 }
 
+const thumbQuality = 75
+
 // GetPostImage returns a signed S3 URL to return an image file
-func (api *API) GetPostImage(ctx context.Context, id uint32, filePath string) (string, error) {
-	bucket, err := api.OpenBucket(ctx)
+func (api *API) GetPostImage(ctx context.Context, id uint32, filePath string, expireSeconds, height, width int) (*ImageMetadata, error) {
+	signingBucket, err := api.OpenBucket(ctx, true)
 	if err != nil {
-		return "", NewError(err)
+		return nil, NewError(err)
+	}
+	defer signingBucket.Close()
+
+	bucket, err := api.OpenBucket(ctx, false)
+	if err != nil {
+		return nil, NewError(err)
 	}
 	defer bucket.Close()
 
 	key := fmt.Sprintf(ImagesPrefix, id) + filePath
-	signedURL, err := bucket.SignedURL(ctx, key, &blob.SignedURLOptions{
-		Expiry: 1 * time.Hour,
+
+	// return full image?
+	if height == 0 && width == 0 {
+		// kind of a shame to read and decode the entire image just to get the dimensions
+		// we read the image once and could store the dimensions in a separate _dimensions.json file for future use,
+		// but this seems like a case of pre-mature optimization
+		// openseadragon, used by the client, needs the image dimensions in order to function properly
```

Comment:
The other option would be to read only the parts of the file that contain size information. That appears to be difficult or impossible to do for JPEGs at least. [jpeg.DecodeConfig](https://golang.org/pkg/image/jpeg/#DecodeConfig) at least doesn't actually decode the image, but it looks like it reads the whole thing. But I agree that this probably is not on the critical path.